src/DynamoCore/Core/DynamoMigrator.cs

```diff
@@ -0,0 +1,251 @@
+
+            try
+            {
+                var serializer = new XmlSerializer(typeof(PreferenceSettings));
+
+                using (var fs = new FileStream(preferenceSettingsFilePath, FileMode.Open, FileAccess.Read))
+                {
+                    settings = serializer.Deserialize(fs) as PreferenceSettings;
+                    fs.Close(); // Release file lock
+                }
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine(ex.Message);
+                Console.WriteLine(ex.StackTrace);
+            }
+            return settings;
+        }
+
+        /// <summary>
+        /// Migrates preference settings and copies packages and custom node 
+        /// definitions from older versions to the currently installed Dynamo version
+        /// </summary>
+        /// <returns></returns>
+        public static PreferenceSettings MigrateBetweenDynamoVersions(string rootFolder)
+        {
+            var assemblyPath = Assembly.GetExecutingAssembly().Location;
+            var currentVersionInfo = FileVersionInfo.GetVersionInfo(assemblyPath);
+
+            // No migration required if the current version is <= version 0.7
+            if (currentVersionInfo.FileMajorPart == 0 &&
+                currentVersionInfo.FileMinorPart <= 7)
+                return null;
+
+            var previousVersion = FindPreviousVersion(rootFolder, currentVersionInfo);
+
+            // No migration required if no previous version exists
+            if (previousVersion.MajorPart == currentVersionInfo.FileMajorPart &&
+                previousVersion.MinorPart == currentVersionInfo.FileMinorPart)
+                return null;
+
+            // Create concrete DynamoMigratorBase object using previousVersion and reflection
+            var sourceMigrator = CreateMigrator(rootFolder, previousVersion.MajorPart, previousVersion.MinorPart);
+            Debug.Assert(sourceMigrator != null);
+            var preferenceSettingsFile = sourceMigrator.ReadPreferences();
+
+            // get migrator object for current version
+            var targetMigrator = CreateMigrator(rootFolder, currentVersionInfo.FileMajorPart, currentVersionInfo.FileMinorPart);
+            Debug.Assert(targetMigrator != null);
+            var preferenceSettings = targetMigrator.WritePreferences(preferenceSettingsFile);
+
+            var sourceDir = Path.Combine(sourceMigrator.CurrentVersionPath, "packages");
+            var targetDir = Path.Combine(targetMigrator.CurrentVersionPath, "packages");
+            Copy(sourceDir, targetDir);
+
+            sourceDir = Path.Combine(sourceMigrator.CurrentVersionPath, "definitions");
+            targetDir = Path.Combine(targetMigrator.CurrentVersionPath, "definitions");
+            Copy(sourceDir, targetDir);
+
+            return preferenceSettings;
+        }
+
+        private static FileVersion FindPreviousVersion(string rootFolder, FileVersionInfo currentVersionInfo)
```

Comment:
Here's the specification for methods:

```csharp
/// <summary>
/// Get a list of file version objects given a root folder. Assuming the 
/// following folders exist:
/// 
///     e:\some\path\0.4
///     e:\some\path\0.75
///     e:\some\path\1.82
/// 
/// Calling this method with "e:\\some\\path" would return an ordered 
/// list in the following way (from largest number to smallest number):
/// 
///     { FileVersion(1, 82), FileVersion(0, 75), FileVersion(0, 4) }
/// 
/// </summary>
/// <param name="rootFolder">The root folder under which versioned sub-
/// folders are expected to be found. This argument must represent a valid
/// local directory.</param>
/// <returns>Return a list of FileVersion objects, ordered by newest 
/// version to the oldest version.</returns>
/// <exception cref="System.ArgumentNullException">This exception is 
/// thrown if rootFolder is null or an empty string.</exception>
/// <exception cref="System.IO.DirectoryNotFoundException">This exception
/// is thrown if rootFolder points to an invalid directory.</exception>
/// 
private static IEnumerable<FileVersion> GetInstalledVersions(string rootFolder)
{
}

```

Let's do one thing:

1. Make a standalone PR for this utility method.
2. Update `FileVersion` so that `List.Sort` can be done.
3. A test case proving this utility work (the test can set up dummy folders in `UnitTestBase.TempFolder` for this).